src/hotspot/share/opto/type.cpp

```diff
@@ -3260,24 +3260,38 @@ void TypeRawPtr::dump2( Dict &d, uint depth, outputStream *st ) const {
 // Convenience common pre-built type.
 const TypeOopPtr *TypeOopPtr::BOTTOM;
 
-TypeInterfaces::TypeInterfaces()
-        : Type(Interfaces), _list(Compile::current()->type_arena(), 0, 0, nullptr),
+TypeInterfaces::TypeInterfaces(const GrowableArray<ciInstanceKlass*>* interfaces)
+        : Type(Interfaces), _interfaces(interfaces),
           _hash(0), _exact_klass(nullptr) {
-  DEBUG_ONLY(_initialized = true);
-}
-
-TypeInterfaces::TypeInterfaces(GrowableArray<ciInstanceKlass*>* interfaces)
-        : Type(Interfaces), _list(Compile::current()->type_arena(), interfaces->length(), 0, nullptr),
-          _hash(0), _exact_klass(nullptr) {
-  for (int i = 0; i < interfaces->length(); i++) {
-    add(interfaces->at(i));
-  }
+  assert(_interfaces != nullptr, "no interfaces should be passed as empty list");
   initialize();
 }
 
-const TypeInterfaces* TypeInterfaces::make(GrowableArray<ciInstanceKlass*>* interfaces) {
-  TypeInterfaces* result = (interfaces == nullptr) ? new TypeInterfaces() : new TypeInterfaces(interfaces);
-  return (const TypeInterfaces*)result->hashcons();
+const TypeInterfaces* TypeInterfaces::make(const GrowableArray<ciInstanceKlass*>* interfaces) {
```

Comment:
I think you can make `_interface` a `ciInstanceKlass**` and do this:

    void* ptr = Type::operator new(sizeof(TypeInterfaces) + sizeof(ciInstanceKlass*) * interfaces->length())

Then `delete ptr` should drop the whole thing.

Reply:
A `GrowableArrayFromArray` would be mostly compatible with the interface of `GrowableArray`, too.

Reply:
Ah! nice. I wasn't aware of `GrowableArrayFromArray`. Updated change follows your suggestion.